segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/StorageHealthTracker.java

```diff
@@ -0,0 +1,316 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.segmentstore.storage.chunklayer;
+
+import com.google.common.util.concurrent.AtomicDouble;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Tracks the health status of the {@link ChunkedSegmentStorage}.
+ */
+@Slf4j
+@RequiredArgsConstructor
+class StorageHealthTracker {
+    /**
+     * Tracks whether storage is slow.
+     */
+    private final AtomicBoolean isStorageDegraded = new AtomicBoolean(false);
```

Comment:
At first glance, this could be removed. I don't see anything in the code that only depends from `isStorageDegraded`. For throttling, we use `isStorageDegraded || shouldThrottle`, and for considering storage unavailable the code uses `isStorageDegraded || isStorageUnavailable`. At the end of the day, we could simplify this code by removing this altogether and leave only `shouldThrottle` and `isStorageUnavailable`, which are self-explanatory of their purpose (`isStorageDegraded` is ambiguous, as it is not clear if it needs to be something else other than throttle and/or cnsidering storage unavailable). As this class has an important function in how data flows (or not) to LTS, the simpler the better.

Reply:
This also helps us to simplify user configuration as well.

Reply:
isStorageDegraded is used to outright reject request instead of just throttling .
But let me see if I can simplify this.

Reply:
What happens to these rejected requests?

Reply:
If they are retried by upper layers (StorageWriter/StorageReader)